File: c++/hard/pashmark_and_parmida.cpp
#include<bits/stdc++.h>
using namespace std;

class BIT{
public:
    vector<int> bit;
    int n;

    BIT(int n){
        this->n=n;
        bit.assign(n+2,0);
    }

    void update(int idx,int val){
        while(idx<=n){
            bit[idx]+=val;
            idx+=idx&(-idx);
        }
    }

    int query(int idx){
        int sum=0;

        while(idx>0){
            sum+=bit[idx];
            idx-=idx&(-idx);
        }
        return sum;
    }
};

int main(){

    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin>>n;

    vector<int>a(n+1);

    for(int i=1;i<=n;i++)
        cin>>a[i];

    vector<int>left(n+1),right(n+1);

    unordered_map<int,int> mp;

    for(int i=1;i<=n;i++)
        left[i]=++mp[a[i]];

    mp.clear();

    for(int i=n;i>=1;i--)
        right[i]=++mp[a[i]];

    BIT bit(n);

    long long ans=0;

    for(int i=n;i>=1;i--){

        ans += bit.query(left[i]-1);

        bit.update(right[i],1);
    }

    cout<<ans;
}

File: js/hard/pashmark-and-parmida.js
const fs=require("fs");

const input=fs.readFileSync(0,"utf8").trim().split(/\s+/);

let p=0;

let n=+input[p++];

let a=[];

for(let i=0;i<n;i++)
    a.push(+input[p++]);

let left=Array(n+1).fill(0);
let right=Array(n+1).fill(0);

let mp=new Map();

for(let i=1;i<=n;i++){

    let x=a[i-1];

    let v=(mp.get(x)||0)+1;

    mp.set(x,v);

    left[i]=v;
}

mp.clear();

for(let i=n;i>=1;i--){

    let x=a[i-1];

    let v=(mp.get(x)||0)+1;

    mp.set(x,v);

    right[i]=v;
}

let bit=Array(n+2).fill(0);

function update(i,val){

    while(i<=n){

        bit[i]+=val;

        i+=i&-i;
    }
}

function query(i){

    let s=0;

    while(i>0){

        s+=bit[i];

        i-=i&-i;
    }

    return s;
}

let ans=0;

for(let i=n;i>=1;i--){

    ans+=query(left[i]-1);

    update(right[i],1);
}

console.log(ans.toString());